modules/client/src/main/java/org/apache/ignite/internal/client/table/ClientTableMapUtils.java

```diff
@@ -0,0 +1,243 @@
+import org.apache.ignite.internal.logger.IgniteLogger;
+import org.apache.ignite.internal.util.CompletableFutures;
+import org.apache.ignite.internal.util.ExceptionUtils;
+import org.apache.ignite.tx.Transaction;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Provides batch map utility methods.
+ */
+class ClientTableMapUtils {
+    // TODO https://issues.apache.org/jira/browse/IGNITE-27073
+    private static final long DEFAULT_IMPLICIT_GET_ALL_TIMEOUT_NANOS = TimeUnit.MILLISECONDS.toNanos(5000);
+
+    static <R, E> void mapAndRetry(
+            MapFunction<E, R> mapFun,
+            @Nullable R initialValue, Reducer<R> reducer,
+            List<Transaction> txns,
+            Map<Integer, List<E>> mapped,
+            long[] startTs,
+            CompletableFuture<R> resFut,
+            IgniteLogger log
+    ) {
+        if (startTs[0] == 0) {
+            startTs[0] = System.nanoTime();
+        }
+
+        List<CompletableFuture<R>> res = new ArrayList<>();
+
+        for (Entry<Integer, List<E>> entry : mapped.entrySet()) {
+            res.add(mapFun.apply(entry.getValue(), PartitionAwarenessProvider.of(entry.getKey()), mapped.size() > 1));
+        }
+
+        CompletableFutures.allOf(res).handle((ignored, err) -> {
+            List<CompletableFuture<Void>> waitCommitFuts = List.of();
+            if (!txns.isEmpty()) {
+                if (err != null) {
+                    boolean needRetry = unlockOnRetry(txns, res, log);
+
+                    long nowRelative = System.nanoTime();
+                    if (needRetry && nowRelative - startTs[0] < DEFAULT_IMPLICIT_GET_ALL_TIMEOUT_NANOS) {
+                        startTs[0] = nowRelative;
+                        txns.clear(); // The collection is re-filled on next map attempt.
+
+                        mapAndRetry(mapFun, initialValue, reducer, txns, mapped, startTs, resFut, log);
+
+                        return null;
+                    }
+
+                    resFut.completeExceptionally(err);
+
+                    return null;
+                }
+
+                waitCommitFuts = unlockFragments(txns, log);
+            } else {
+                if (err != null) {
+                    resFut.completeExceptionally(err);
+
+                    return null;
+                }
+            }
+
+            R in = initialValue;
+
+            for (CompletableFuture<R> val : res) {
+                in = reducer.reduce(in, val.getNow(null));
+            }
+
+            if (waitCommitFuts.isEmpty()) {
+                resFut.complete(in);
+            } else {
+                R finalIn = in;
+                CompletableFutures.allOf(waitCommitFuts).whenComplete((r, e) -> {
+                    // Ignore errors.
+                    resFut.complete(finalIn);
+                });
+            }
+
+            return null;
+        });
+    }
+
+    static <E> void mapAndRetry(
```

Comment:
This method looks identical to the above except for the reducer, can we extract the common part?